### ISSUE TITLE
API / Webhook errors

### ISSUE THREAD

**Issue by jennyCognio** [reapit/foundations#6651]
We've had approx 2200 errors through the API today in two batches

~1800  between 9:07 and 9:45.
~ 400 at 10:30

Contact  created / updated webhook - error fetching records from Reapit 

These were reported back to us as a combination of  500 and 429 errors which seems to point to it being your rate limiting again (with your system sending us more webhooks that you can handle getting back as API requests for the data)

I was told to add this as a bug so that we can find out what was being updated to cause this volume of calls - if it's something that affects the data we need to display on the website we'll need to manually run in these contacts again; hopefully it isn't.

Thanks

**Comment by jennyCognio** [comment]
Hi
Can this be escalated please - we've had another 3500 errors today and they're still happening

**Comment by plittlewood-rpt** [comment]
Hi @jennyCognio - the rate at which you receive webhooks could quite easily be greater than the rate at which you are permitted to hit our API yourself. It is down to the architecture of your own application to ensure you do not exceed the rate limits enforced on the Platform. One thing I'd question here is why you are using webhooks if you are then going off to the API to fetch the data anyway. The data you need is in the webhook payload so it almost sounds like you are doubling up on what's needed here. Can you elaborate a bit on your application workflow please?

**Comment by jennyCognio** [comment]
Hi Pete
Are you saying that your system can fire webhooks to our application at a
speed that you can't handle coming back to your own API - have I
misunderstood?
Unfortunately the webhooks do not contain what is needed as we require
related data to update our records and the webhooks themselves have minimal
information - this was fully specced out with the team at Reapit when we
first started our implementation.

Whilst this has only very occasionally been an issue over the last 9 months
(and one which we could fix from this end) over the last 72 hours we've had
tens of thousands of these errors coming back from you - this ticket isn't
because we expect you to be able to improve the API rate, just to find out
why the situation has degraded so much in the last few days and to get that
fixed please.

Would this be easier to address/figure out in a phone call?


Best wishes
Jenny

**Comment by plittlewood-rpt** [comment]
Hi @jennyCognio - the webhooks system is separate to the Platform APIs and runs on different infrastructure. We have to be able to get the notifications out as quickly as is possible to avoid any backlogs forming during busy periods. It is up to your application to queue work as appropriate to ensure you do not exceed our rate limits. I will ensure we look into what might be causing the apparent increase in these issues however it's worth noting we have made some improvements to our webhooks pipeline recently that will mean there is less latency between an event occurring in AgencyCloud and the webhooks being sent. As a result you'll be seeing a higher/faster throughput. As webhooks are driven entirely by the database, another option is that some other actions are taking place on the customer database(s) you are subscribed to webhooks. As you'll appreciate we can't really control end user behaviour. I'll check with our service desk to see if there's been any maintenance jobs taking place that might be affecting this also. Is there a particular customer you are seeing problems with? It would be great if you could provide some information about the customers affected.

**Comment by jennyCognio** [comment]
Hi Pete
I don't think this is anything to do with latency as it's about the volume
increase causing you guys issues at your end, not the speed at which a
webhook is sent.
We should only be getting notified when a record is changed - there should
be no reason why we're getting more webhooks as the number of records being
updated shouldn't have changed in the last week.

(unless you're sending more than one webhook for each update - which in
turn could relate to a previous bug ticket we had last year about getting
multiple calls for each change - this was ticket #5229 and it hasn't been
updated since 7 Jan when we were told you were looking into it further)

Best wishes
Jenny

Jenny Forster



Cognio Digital Ltd      12 Melcombe place, Marylebone, London, NW1 6JJ   Tel:
0800 433 7045
 www.cogniodigital.com
<https://mailtrack.io/trace/link/7ca898632c970fee285c5542ec4d38c7d6d81880?url=http%3A%2F%2Fwww.cogniodigital.com&userId=2673365&signature=38ec8e41e96dc8cd>
   Company Reg: 11227933   VAT No: 290277882

**Comment by plittlewood-rpt** [comment]
Hi @jennyCognio - thanks for your feedback. Please can you confirm which customer(s) you are being affected by? I'm assuming at this point it's BEF but if you can confirm that would be great.

**Comment by jennyCognio** [comment]
Hi
Yes, it's BEF
Best wishes
Jenny

Jenny Forster



Cognio Digital Ltd      12 Melcombe place, Marylebone, London, NW1 6JJ   Tel:
0800 433 7045
 www.cogniodigital.com
<https://mailtrack.io/trace/link/6d543fe0de81594454f0e6386b2ae7f8f9e8153f?url=http%3A%2F%2Fwww.cogniodigital.com&userId=2673365&signature=d72dba3fb01a10f2>
   Company Reg: 11227933   VAT No: 290277882

**Comment by plittlewood-rpt** [comment]
Hi @jennyCognio - I have had a look at the BEF dataset and can see a huge number of updates in their system on 20th April in the morning. I will check in with our service team to see what maintenance was being done so we can attribute the cause. Ultimately though, as you are subscribed to webhook events on various entities, any kind of batch job (either done by us or done by a client) _will_ result in you being notified otherwise your copy of the record would be out of date.

**Comment by plittlewood-rpt** [comment]
Hi @jennyCognio I have got to the bottom of what caused this. A number of negotiator data transfers were carried out on the morning of 20th April. This action typically happens when somebody leaves an agency business and their data gets transferred to a colleague who picks up their contacts etc. There were at least 3 negotiator transfers done on this day which results in the negotiatorId(s) being updated on various records. As a result of these updates, webhooks get emitted to notify any listeners of the changes.

**Comment by jennyCognio** [comment]
Thanks Pete
Do these get delayed and sent out over multiple days? Because we were still
getting thousands of errors back from your API yesterday.



Best wishes
Jenny

Cognio Digital

**Comment by plittlewood-rpt** [comment]
@jennyCognio no they all get sent as they happen over a period of a x minutes (x is a sliding scale depending on the volume going through at any one time but we scale according to demand). Webhooks are sent in as near real-time as possible. It's up to the listener to make sure they don't exceed rate limits if they choose to call back into the API upon receipt of a webhook.  If you were getting loads through yesterday as well it's likely there were similar updates yesterday too. This is something I can check

**Comment by jennyCognio** [comment]
Thanks Pete
They seem to have stopped now but it would be good if there was some way we
could find out what you are updating when you do this please - even if it's
just a courtesy email of "we're updating X over the next couple of days"
that way we can decide if the resulting errors mean we need to run in the
records again when your system isn't as busy, or if they can be left.
Just because I'll be asked, when you do these auto updates, the resultant
API calls aren't charged are they?
Thanks

Best wishes
Jenny

Jenny Forster



Cognio Digital Ltd      12 Melcombe place, Marylebone, London, NW1 6JJ   Tel:
0800 433 7045
 www.cogniodigital.com
<https://mailtrack.io/trace/link/84a84ebe11f50979c8e2b5e3675644dd74d468e3?url=http%3A%2F%2Fwww.cogniodigital.com&userId=2673365&signature=73597ec27c464758>
   Company Reg: 11227933   VAT No: 290277882

**Comment by HollyJoyPhillips** [comment]
Hi @jennyCognio, Unfortunately, those updates are not controlled by Platform. We process the data as and when it occurs. If you want to setup a notification, I would suggest asking Beresford’s to raise a ticket to service desk.

However, it is most likely that those requests (to transfer data from one neg to another) would have originated from Berfesfords (to the service desk) or even carried out by an Admin by Beresfords themselves in the office (if they have required permissions/correct setup). Might be worth verifying as you'll be able to get a notification from Beresford's direct. 

The webhooks are setup to emit events based on events that occur in the AgencyCloud DB. Either through the UI by the customer or on a Database level by a service desk representative. Either way, it is not something we can help within the Platform department.

All webhooks calls are charged per API call, regardless of the event type.

**Comment by plittlewood-rpt** [comment]
Closing item as behaviour has been investigated and explained.

**Comment by jennyCognio** [comment]
Hi Pete
We asked the client to let us know when they were transferring data from one neg to another. This happened again yesterday and there were no timeout errors from you. 

Taking this and the sheer number of errors there were, over multiple days, it does suggest that updating a neg is not the cause of the volume of webhooks sent. Could you look and see if you were doing anything internally please? Unfortunately this has had quite a few knock on effects with data and we need to make a judgement call on our next steps, which we can only do once we know what caused it and if it's going to be a frequent occurrence.

Thanks
Jenny

**Comment by jennyCognio** [comment]
Hi Pete
My apologies, I meant rate limited, not timing out.

We understand your rate limiting as a general issue, but normally we don't
have too many errors, maybe 20-30 a day which are manageable through a
manual update.

The issue is the circa 16k errors between 20 April and 28 April

To give you context, we have had around 10k errors from launch to 20April,
including several thousand at/after launch, and 3k on 28/10/21.

Since 28 April. errors have reverted to the 'normal' level

We don't think transferring three negs on the 20 April would have caused
16k errors over 8 days - to allow us to accurately assess the risk going
forward, we would really like to know what happened and whether it could
happen again.

Thanks

Best wishes
Jenny

**Comment by HollyJoyPhillips** [comment]
Hi @jennyCognio, 

As mentioned above, the amount of webhooks that were issued on the 20th was the result of a data transfer by an ‘S Craske’ for 4 negotiators. This amounted to around 4000 + contact records which in turn have associated records, vendor, applicant, property, and tenancies which explains the 10k webhooks.

The data is aligned to the amount of webhooks emitted.

Looking back on the thread, I can’t see an example the error that you are receiving. It might be that we probably need to explore that seeing as the data is accurate.

If you can provide an example, we can investigate further.
Thanks
Holly

**Comment by jennyCognio** [comment]
Hi Holly
Are you confirming that the volume of errors received over the 8 days is as
expected? The transfers were made on a single day (according to both Reapit
and our client) but the errors (API rate limiting) were received in their
thousands for 8 days following that.

From our end, the webhooks are received and the API call is immediately
made. As you've said you don't queue at your end, I don't see how this
could have triggered rate limiting errors 8 days later. We're just trying
to figure out what did, and if it's going to be a frequent occurrence.

Happy to have a call if easier to explain?

Thanks
Jenny

**Comment by jennyCognio** [comment]
Hi Holly
I needed to check with the client that they weren't discussing this separately in other conversations but they've asked me to continue it here.
If you look at the first few messages in the ticket you'll see the issues that were having here aren't specifically to do with your rate limiting, but about the 16k errors we got due to it between the dates 20-28April.

We understand the rate limiting and usually it's not a big issue. Between these dates however, it was, as the volume of webhooks received vastly outstripped anything we usually get and meant that all our data got out of sync.

What we're trying to establish is what caused this spike in webhooks. If it was just on the 20th April it could have theoretically be down to transferring the data from one neg to another as this happened 3-4 times on that date.

However, Pete has confirmed that those webhooks would have been sent as near to immediately as possible. So any interruption would have been limited to an hour or so around the time of the transfer.

Instead, we had almost 8 days of thousands of errors a day, and then it stopped and we're back to our normal level.

Additionally, when we made another neg transfer on the 9th May and no errors were received.

We're trying to establish what happened between 20-28th April and if it could happen again (and if so, when) to help us make a decision about how much additional development might be required and when.

Thanks
Jenny
